src/symbol/collision_index.js

```diff
@@ -143,99 +124,123 @@ class CollisionIndex {
-            if (atLeastOneCirclePlaced) {
-                const dx = projectedPoint.x - placedCollisionCircles[placedCollisionCircles.length - 4];
-                const dy = projectedPoint.y - placedCollisionCircles[placedCollisionCircles.length - 3];
-                // The circle edges touch when the distance between their centers is 2x the radius
-                // When the distance is 1x the radius, they're doubled up, and we could remove
-                // every other circle while keeping them all in touch.
-                // We actually start removing circles when the distance is √2x the radius:
-                //  thinning the number of circles as much as possible is a major performance win,
-                //  and the small gaps introduced don't make a very noticeable difference.
-                const placedTooDensely = radius * radius * 2 > dx * dx + dy * dy;
-                if (placedTooDensely) {
-                    const atLeastOneMoreCircle = (k + 8) < collisionCircles.length;
-                    if (atLeastOneMoreCircle) {
-                        const nextBoxDistanceToAnchor = collisionCircles[k + 8];
-                        if ((nextBoxDistanceToAnchor > -firstTileDistance) &&
-                        (nextBoxDistanceToAnchor < lastTileDistance)) {
-                            // Hide significantly overlapping circles, unless this is the last one we can
-                            // use, in which case we want to keep it in place even if it's tightly packed
-                            // with the one before it.
-                            markCollisionCircleUsed(collisionCircles, k, false);
-                            continue;
-                        }
-                    }
+            let segments = [];
+
+            if (projectedPath.length > 0) {
+                // Quickly check if the path is fully inside or outside of the padded collision region.
+                // For overlapping paths we'll only create collision circles for the visible segments
+                const minPoint = projectedPath[0].clone();
+                const maxPoint = projectedPath[1].clone();
+
+                for (let i = 1; i < projectedPath.length; i++) {
+                    minPoint.x = Math.min(minPoint.x, projectedPath[i].x);
+                    minPoint.y = Math.min(minPoint.y, projectedPath[i].y);
+                    maxPoint.x = Math.max(maxPoint.x, projectedPath[i].x);
+                    maxPoint.y = Math.max(maxPoint.y, projectedPath[i].y);
+                }
+
+                const region = computeRectClipRegion(minPoint, maxPoint, screenPlaneMin, screenPlaneMax);
+
+                // 0 == fully visible, -1 == partly visible, > 1 not visble at all
+                if (region === 0) {
+                    segments = [projectedPath];
+                } else if (region < 0) {
+                    segments = clipLine([projectedPath], screenPlaneMin.x, screenPlaneMin.x, screenPlaneMax.x, screenPlaneMax.y);
```

Comment:
This isn't very important but I'm wondering if it's worth clipping here. I haven't profiled but I'm wondering if it would be just as cheap to generate circles for the parts of the lines that are offscreen and do bounds-checks on circles individually.

Reply:
In most scenarios this would be true. Problems arise with overscaled tiles where labels can be near the front plane. A point near the front plane (distance < 0.1) can be projected really far away from the screen (due to the division by `w`). I've encountered cases where more than 1000 circles were generated because one vertex of the line was near the front plane.

Reply:
Yep, that makes sense! Thanks